### PR TITLE
Add unit tests for qos and qos_event files

### PR DESCRIPTION
**Pull request by brawner**
This PR add coverage for the remaining lines and functions in qos.hpp, qos.cpp, qos_event.hpp, qos_event.cpp

Signed-off-by: Stephen Brawner <brawner@gmail.com>

### CODE REVIEW

**Comment by brawner** [comment]
Force pushed some test fixes. It turns I had qos_event in my ctest exclude pattern because it ran for 10 seconds with the fastrtps. I added a check in the offending test, so now rclcpp's tests run 10 seconds faster!

File: rclcpp/test/rclcpp/test_qos_event.cpp [hunk 1]
```diff
@@ -207,7 +209,9 @@ TEST_F(TestQosEvent, test_default_incompatible_qos_callbacks)
   rclcpp::executors::SingleThreadedExecutor ex;
   ex.add_node(node->get_node_base_interface());
 
-  ex.spin_until_future_complete(log_msgs_future, std::chrono::seconds(10));
+  // This future won't complete on fastrtps, so just timeout immediately
+  const auto timeout = (is_fastrtps) ? std::chrono::milliseconds(5) : std::chrono::seconds(10);
+  ex.spin_until_future_complete(log_msgs_future, timeout);
```
Comment (clalancette) [inline rclcpp/test/rclcpp/test_qos_event.cpp:214]: Thinking about this test more, can't we just skip it completely?  That is, at the top of this method, just have:

```
if (is_fastrtps) {
  return;
}
```

**Review by clalancette — COMMENTED** [review]
I've got one thing to improve, but awesome that you saved 10 seconds on Fast-RTPS!

File: rclcpp/test/rclcpp/test_qos_event.cpp (hunk 1 above)
Comment (brawner) [reply rclcpp/test/rclcpp/test_qos_event.cpp:214]: gtest does have a macro for runtime skipping of tests (`GTEST_SKIP`), but that would make it harder for this issue to be detected in the future if the code changes for fastrtps.

I updated the test to just check that the logger messages are empty.

**Comment by brawner** [comment]
Testing `--packages-select rclcpp`

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12546/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7514/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10257/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12466/)

**Review by clalancette — APPROVED** [review]
The code looks good to me, so I'll approve.  Windows is unhappy, though.

File: rclcpp/test/rclcpp/test_qos_event.cpp [hunk 2]
```diff
@@ -222,3 +229,97 @@ TEST_F(TestQosEvent, test_default_incompatible_qos_callbacks)
...
+          callback, rcl_publisher_event_init, rcl_handle, event_type).reset();
+      };
+    // This is done through a lambda because the compiler is having trouble parsing the templated
+    // function inside a macro.
+    EXPECT_THROW(throwing_statement(), rclcpp::exceptions::RCLError);
+  }
+
+  {
+    // Logs error and returns
+    auto mock = mocking_utils::patch_and_return(
```
Comment (Blast545) [inline rclcpp/test/rclcpp/test_qos_event.cpp:260]: Why not `inject_on_return` as it is a `fini` method?

**Review by Blast545 — APPROVED** [review]
LGTM, minor suggestion

File: rclcpp/test/rclcpp/test_qos_event.cpp (hunk 2 above)
Comment (brawner) [reply rclcpp/test/rclcpp/test_qos_event.cpp]: Done

**Comment by brawner** [comment]
I think I resolved the windows issue.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12567/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7535/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10278/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12490/)

**Comment by brawner** [comment]
Trying windows again:
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12495/)

File: rclcpp/test/rclcpp/test_qos_event.cpp [hunk 3]
```diff
@@ -222,3 +229,97 @@ TEST_F(TestQosEvent, test_default_incompatible_qos_callbacks)
...
+      "lib:rclcpp", rcl_publisher_event_init, RCL_RET_ERROR);
+
+    auto throwing_statement = [callback, rcl_handle, event_type]() {
+        // reset() is not needed for the exception, but it handles unused return value warning
+        std::make_shared<
+          rclcpp::QOSEventHandler<decltype(callback), std::shared_ptr<rcl_publisher_t>>>(
```
Comment (brawner) [inline rclcpp/test/rclcpp/test_qos_event.cpp:250]: @clalancette Here's a situation where using `auto` shot me in the foot. For some reason with windows, capturing rcl_handle by reference in the lambda and then calling decltype(rcl_handle) caused it to think that this should have been a `shared_ptr<rcl_publisher_t>&`, which then caused it to complain about an uninitialized reference of all things.

**Comment by brawner** [comment]
I just fixed and ran things on windows for the latest changes. Everything passed successfully before so merging this as is.

**Pull request merged**